ord_schema/message_helpers.py

```diff
@@ -782,3 +787,96 @@ def create_message(message_name):
     except (AttributeError, TypeError) as error:
         raise ValueError(
             f'Cannot resolve message name {message_name}') from error
+
+
+def messages_to_dataframe(messages: Iterable[any_pb2.Any],
+                          drop_constant_columns: bool = False) -> pd.DataFrame:
+    """Converts a list of protos to a pandas DataFrame.
+
+    Args:
+        messages: List of protos.
+        drop_constant_columns: Whether to drop columns that have the same value
+            for all rows.
+
+    Returns:
+        DataFrame.
+    """
+    rows = []
+    for message in messages:
+        rows.append(message_to_row(message))
+    df = pd.DataFrame(rows)
+    if drop_constant_columns:
+        drop = []
+        for column in df.columns:
+            if len(df[column].unique()) == 1:
+                drop.append(column)
+        for column in drop:
+            del df[column]
+    return df
+
+
+def message_to_row(message: any_pb2.Any,
+                   trace: Tuple[str] = None) -> Dict[str, ScalarType]:
+    """Converts a proto into a flat dictionary mapping fields to values.
+
+    The keys indicate any nesting; for instance a proto that looks like this:
+
+    value: {
+      subvalue: 5
+    }
+
+    will show up as {'value.subvalue': 5} in the dict.
+
+    Args:
+        message: Proto to convert.
+        trace: Tuple of strings; the trace of nested field names.
+
+    Returns:
+        Dict mapping string field names to scalar value types.
+    """
+    if trace is None:
+        trace = tuple()
+    row = {}
+    for field, value in message.ListFields():
+        if field.label == field.LABEL_REPEATED:
+            if (field.type == field.TYPE_MESSAGE and
+                    field.message_type.GetOptions().map_entry):
+                value_field = field.message_type.fields_by_name['value']
+                for key, subvalue in value.items():
+                    this_trace = trace + (f'{field.name}["{key}"]',)
+                    row.update(
```

Comment:
Should we explicitly check for whether a field in the dict will be overwritten? Based on your naming system, I don't see why it would be, but wanted to throw that out as an idea.

Reply:
Yes, this is a good check. Done.